decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+package openlr.decoder.backtracking;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BackTrackingGraph<T, L> {
+    private final List<Slice<T, L>> slices;
+    private final Map<String, GraphEdge> edges;
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T, L> graphEdge = null;
+    private GraphEdge<T, L> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice<T, L>> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    /**
+     * if the user has marked last edge returned by the graph as red;
+     * then the next edge should be an unmarked edge with the same start node
+     * But If all the edges starts from this start node is already marked
```

Comment:
But **if** all the edges **starting** from this start node **are** already marked

Reply:
31890348b9aef835f9f9065ddd8625de1cfab0ea